### PR TITLE
xbits: error out on unaccepted values

### PR DESCRIPTION
**Pull request by inashivb**
If we have unaccepted values as a part of xbits keyword, they are
silently passed when it should fail. This patch compares the total
length of the raw string with the match patterns and errors out if
anything apart from the matched patterns is present.

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/4786

### CODE REVIEW

File: src/detect-xbits.c [hunk 1]
```diff
@@ -271,6 +275,33 @@ static int DetectXbitParse(DetectEngineCtx *de_ctx,
...
+                 * up to 2n+1 values, where n = no. of capturing parentheses */
+                if (bufptr != prev_buf) { // FIXME Dangerous check? Could it be wrong in any cases?
+                    args_len += strlen(bufptr);
+                }
+                prev_buf = bufptr;
+            }
+            if (args_len != stripped_rawstr_len) {
+                SCLogError(SC_ERR_INVALID_VALUE, "Invalid values found in the xbits setting");
+                return -1;
+            }
```
Comment (inashivb) [inline src/detect-xbits.c:304]: This is probably too long and seemingly dirty way to achieve the fix. Better suggestions?

File: src/detect-xbits.c [hunk 2]
```diff
@@ -546,6 +577,87 @@ static int XBitsTestSig02(void)
...
+    Signature *s = DetectEngineAppendSig(de_ctx,
+            "drop http any any -> any any (msg:\"Malicious file detected\"; \
+            filesha256:blacklist; xbits:set, blocked_http, track ip_pair, \
+            expire 3600; sid:2; rev:2;");
+    FAIL_IF_NULL(s);
+
+    DetectEngineCtxFree(de_ctx);
+    XBitsTestShutdown();
+    PASS;
+}
```
Comment (inashivb) [inline src/detect-xbits.c:658]: Found this test throw a SEGV in s-v but nowhere in the xbits code, during file hash processing. Gotta know better.

**Comment by inashivb** [comment]
OP says on the ticket
> I suggest adding a bit of context which indicates it should be a standalone keyword in the rule and not an "option" to the xbits keyword.

but I see https://github.com/OISF/suricata/blob/master/src/detect-xbits.c#L277 I _think_ this came from `hostbits` implementation but I'm not sure if it should or should not be an acceptable command for `xbits`?

**Comment by suricata-qa** [comment]
ERROR: ASAN TEST FAIL in tlpr1_asan_cfg QA test

ERROR: QA failed on tlpr1_asan_cfg.
  

Pipeline 5613

**Comment by victorjulien** [comment]
I think especially for the git master I would like to get rid of the pcre use here, and use a simpler tokenizer instead.

**Comment by victorjulien** [comment]
> I think especially for the git master I would like to get rid of the pcre use here, and use a simpler tokenizer instead.

For example like in `DetectDatasetParse` and `DetectMetadataParse`

**Comment by inashivb** [comment]
Replaced w #6905

**Pull request closed without merging**